### PR TITLE
JDK-8305507 Add support for grace period before AbortVMOnSafepointTimeout triggers

### PR DESCRIPTION
**Pull request by w-kudla**
As described https://bugs.openjdk.org/browse/JDK-8305507 this change adds a grace period before AbortVMOnSafepointTimeout activates to allow for JVM warm-up to conclude

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305507](https://bugs.openjdk.org/browse/JDK-8305507): Add support for grace period before AbortVMOnSafepointTimeout triggers (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) ⚠️ Review applies to [9649c3fe](https://git.openjdk.org/jdk/pull/13386/files/9649c3fe40f992d832a49bce52be107df62e0f40)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13386/head:pull/13386` \
`$ git checkout pull/13386`

Update a local copy of the PR: \
`$ git checkout pull/13386` \
`$ git pull https://git.openjdk.org/jdk.git pull/13386/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13386`

View PR using the GUI difftool: \
`$ git pr show -t 13386`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13386.diff">https://git.openjdk.org/jdk/pull/13386.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13386#issuecomment-1517580013)

### CODE REVIEW

File: src/hotspot/share/runtime/safepoint.cpp [hunk 1]
```diff
@@ -784,7 +785,7 @@ void SafepointSynchronize::print_safepoint_timeout() {
 
   // To debug the long safepoint, specify both AbortVMOnSafepointTimeout &
   // ShowMessageBoxOnError.
-  if (AbortVMOnSafepointTimeout) {
+  if (AbortVMOnSafepointTimeout && Management::ticks_to_ms(os::elapsed_counter()) > (jlong)AbortVMOnSafepointTimeoutDelay) {
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/safepoint.cpp:788]: Rather than use the Management API here perhaps just:
```
if (AbortVMOnSafepointTimeout
    && (os::elapsedTime() * MILLIUNITS > AbortVMOnSafepointTimeoutDelay)) {
```
?

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
The request seems reasonable to accommodate, but please see below.

Please add a new testcase for this to test/hotspot/jtreg/runtime/Safepoint/TestAbortVMOnSafepointTimeout.java

Thanks

File: src/hotspot/share/runtime/safepoint.cpp (hunk 1 above)
Comment (w-kudla) [reply src/hotspot/share/runtime/safepoint.cpp]: To my understanding `os::elapsedTime() * MILLIUNITS` will reduce the granularity of the grace period down to a second but I fully agree this is nicer and we don't need millis for this functionality. 
Thank you for the suggestion, @dholmes-ora

**Comment by w-kudla** [comment]
> Please add a new testcase for this to test/hotspot/jtreg/runtime/Safepoint/TestAbortVMOnSafepointTimeout.java

Test case added

File: src/hotspot/share/runtime/safepoint.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/safepoint.cpp]: `os::elapsedTime()` returns a double so has fractional seconds.

File: test/hotspot/jtreg/runtime/Safepoint/TestAbortVMOnSafepointTimeout.java [hunk 2]
```diff
@@ -56,6 +57,31 @@ public static void main(String[] args) throws Exception {
...
+                "-XX:SafepointTimeoutDelay=50",
+                "-XX:GuaranteedSafepointInterval=1",
+                "-XX:-CreateCoredumpOnCrash",
+                "-Xms64m",
+                "TestAbortVMOnSafepointTimeout$TestWithDelay"
+        );
+
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.shouldContain(TestWithDelay.PRE_STALL_TEXT);
+        verifyAbortVmApplied(output);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Safepoint/TestAbortVMOnSafepointTimeout.java:81]: Surely in this case there is no abort ??? Otherwise how are you actually testing the delay?

**Comment by pete23hsbc** [comment]
Hi, unfortunately @w-kudla is out sick. If he's not back this week I'll pick this up for HSBC and look at the test changes requested above.

File: test/hotspot/jtreg/runtime/Safepoint/TestAbortVMOnSafepointTimeout.java (hunk 2 above)
Comment (pete23hsbc) [reply test/hotspot/jtreg/runtime/Safepoint/TestAbortVMOnSafepointTimeout.java:81]: @dholmes-ora I've read through the Test and think I now understand it, don't believe it needs a change but maybe I've missed something.

Process is started with 1500ms arming delay and 50ms safepoint timeout, with abort on timeout set.

A 999ms pause is induced. This would trigger the timeout and hence abort, but we are still within the arming delay window. We output PRE_STALL_TEXT so that the test can verify the arming delay succeeded. (tested on line 80).

A further 999ms pause is induced. This will take us past the end of the arming delay and therefore should trigger the VM abort. This is why we still expect the VM to abort (tested line 81).

tl;dr The test is checking that the abort timeout is initially disarmed, and also that it is correctly rearmed, hence why we still get an abort.

Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/Safepoint/TestAbortVMOnSafepointTimeout.java:81]: @pete23hsbc thank you for that explanation. I see now how the test is expecting to work.

**Review by dholmes-ora — APPROVED** [review]
This seems fine to me.

Thanks.

**Comment by w-kudla** [comment]
Hi @dholmes-ora do you think we can integrate this PR?

**Comment by dholmes-ora** [comment]
@w-kudla  a second review is required before this can be integrated (as per normal hotspot rules). I will see if I can get someone to do it.

File: test/hotspot/jtreg/runtime/Safepoint/TestAbortVMOnSafepointTimeout.java [hunk 3]
```diff
@@ -78,4 +112,21 @@ public static void main(String[] args) throws Exception {
...
     }
+
+    public static class TestWithDelay {
+
+        public static final String PRE_STALL_TEXT = "THE FOLLOWING STALL SHOULD BE CAPTURED";
+
+        public static void main(String[] args) throws Exception {
+            WhiteBox wb = WhiteBox.getWhiteBox();
+            // induce a stall that should not be picked up before grace period
+            wb.waitUnsafe(999);
```
Comment (pchilano) [inline test/hotspot/jtreg/runtime/Safepoint/TestAbortVMOnSafepointTimeout.java:123]: So if we reach this line after 500ms from the time we set the initial counter in os::init() we will trigger the abort here instead. I measured that time just to check how close we are and in my testing machine I see we get here at about 130ms so probably is good enough, but maybe we should add 1 second to AbortVMOnSafepointTimeoutDelay just to be safe?

**Review by pchilano — APPROVED** [review]
Looks good.

**Comment by dholmes-ora** [comment]
@w-kudla please respond to @pchilano suggestion above. Thanks

**Comment by w-kudla** [comment]
> @w-kudla please respond to @pchilano suggestion above. Thanks

No problem, @dholmes-ora . I'm back from holiday next week and will update this then

**Comment by w-kudla** [comment]
Hey @pchilano , I just increased the pre-abort period slightly. Does this work for you?

**Comment by pchilano** [comment]
> Hey @pchilano , I just increased the pre-abort period slightly. Does this work for you?
>
Thanks @w-kudla. It's not ideal to have this dependence on timing but I don't think we can avoid it to test this flag. I guess testing will tell if the window is enough.

**Review by dholmes-ora — APPROVED** [review]
Still good. Thanks

**Comment by w-kudla** [comment]
Can we please get this sponsored? @dholmes-ora @pchilano
